### PR TITLE
Add android support

### PR DESCRIPTION
**Pull request by Beyley**
Android support, moved from #24 due to inactivity from daCyuubi

### CODE REVIEW

**Comment by Beyley** [comment]
@peppy It seems the osu!stream site is not sending most of the previews correctly, only the "Unreleased Tracks"'s previews actually send me any data, the rest send me an empty array for some reason, if its something on my end im doing wrong, please give me a hand because im kinda stuck there, but im leaning towards it possibly being a server issue

**Comment by Beyley** [comment]
I got the resolution issue fixed fully, all i need to do now is figure out how to fix the most unforseen issue for me, my display is too wide, over 2x wider then tall, so a lot of graphics just *cut off*

**Comment by Beyley** [comment]
heres the broken stuff

**Comment by Beyley** [comment]
I cant find the original files for the headers so i cant expand them, would it be acceptable to stretch the size of the broken assets to the width of the window if the screen is too big? (this does not mean squishing them if the screen is too small)

**Comment by peppy** [comment]
If you can provide the specs for your device (or the widest android device) I can make them wider. I've already had to do this three times so far....

**Comment by Beyley** [comment]
The widest i could find is 2.222x aspect ratio (3200x1440), which is the S21, the phone i use, altho there might be some slightly wider ones so maybe 2.3x aspect ratio (3312x1440)?

**Comment by Beyley** [comment]
im p sure that phones will never go over 2.3x so unless this https://www.youtube.com/watch?v=XNBP18nrRdw becomes a real thing, i think we will be fine for the future lol

**Comment by Beyley** [comment]
NVM OKAY, COMPANIES ARE INSANE THERE IS A PHONE WITH A 2.44x ASPECT RATIO?????? https://www.gsmarena.com/motorola_razr_5g-10410.php

maybe go up to 2.5x lol

File: osu!stream/Main.cs [hunk 1]
```diff
@@ -48,7 +50,16 @@ public Application()
...
-            return;
+            if (Director.IsTransitioning) return;
+            
+            if(Director.CurrentOsuMode == OsuMode.Play)
+                if (!((Player)Director.CurrentMode).IsPaused)
+                    (Director.CurrentMode as Player)?.Pause();
+                else if(!(bool)(Director.CurrentMode as Player)?.menu.Failed)
+                    (Director.CurrentMode as Player)?.Resume();
+
+            if (Director.CurrentOsuMode == OsuMode.MainMenu)
```
Comment (peppy) [inline osu!stream/Main.cs:61]: Is this really a thing? I'm pretty sure it shouldn't be.

Comment (Beyley) [reply osu!stream/Main.cs:61]: I tested it and it seems to work as intended, pausing when you press, unpausing when you press again, and not letting you unpause when you fail, i can double check again and record a vid if nessesary

Comment (Beyley) [reply osu!stream/Main.cs:61]: nvm i guess it didnt work fully and i was blind, resume doesnt seem to work properly, ill just press the continue button in the menu automatically

Comment (Beyley) [reply osu!stream/Main.cs]: Resolved in 5256813

Comment (Beyley) [reply osu!stream/Main.cs]: https://youtu.be/hlj7Lm1VJho
video here

Comment (peppy) [reply osu!stream/Main.cs]: The issue is that a mobile app should never be ending the process. Especially on iOS. On android there should be a specific call to handle this that isn't "end execution", I think.

Comment (Beyley) [reply osu!stream/Main.cs]: i pulled it from here, but it seems there might be a better way https://stackoverflow.com/questions/29257929/how-to-terminate-a-xamarin-application

Comment (Beyley) [reply osu!stream/Main.cs]: and also this code is android-only, so it shouldnt be in effect on iOS

Comment (peppy) [reply osu!stream/Main.cs]: What is listed on that page looks more correct. It means that application won't be unloaded from memory, which is what you want.

Comment (Beyley) [reply osu!stream/Main.cs]: The closest things that dont cause issues with assets unloading are `System.Diagnostics.Process.GetCurrentProcess().CloseMainWindow();` which just closes the main window instead of stopping execution (which is p much the same thing cus we reinitialize gamebase on re-load anyway)

Comment (Beyley) [reply osu!stream/Main.cs]: I wish i knew what the home button did so i could just do the same with the back button here but i have no idea how

Comment (peppy) [reply osu!stream/Main.cs]: If this is intended to fix asset (vRAM) reloading then you will need to force quit the process. But I thought this was resolved via #26?

Comment (Beyley) [reply osu!stream/Main.cs]: Pressing the home button then going back was resolved via #26, but it seems that pressing the home button and pressing the back button do different things, so i had the back button disabled, when i dont disable the back button it seems to unload things in a way that even after force reloading *all* assets, everything is a white box, if i can replicate the home button behavior under the back button, it should fix it, and i do eventually want to allow the game to pause and continue properly when you minimize the app, so having one common "minimize" would be nice

Comment (Beyley) [reply osu!stream/Main.cs]: After a LOT of fiddling with stuff i got live tabbing in and out to work (turns out the engine can support it, i just have to hand hold it a bit with texture loading/unloading, which was easy enough)

**Comment by LukaTV939** [comment]
The back button works in-game fine, but can you make the back button usable in the menu like in osu!laser (the latest one)

**Comment by Beyley** [comment]
> The back button works in-game fine, but can you make the back button usable in the menu like in osu!laser (the latest one)?
> Also, how did you import those beatmaps?

Having the back button be usable in the menu is out of scope for this PR, and the maps are from the store

**Comment by Beyley** [comment]
After testing some things it seems to actually break Desktop builds since we never actually properly separated AudioTrackBass from Android/Desktop, since im not on Windows and it dosent build under Mono for some reason, i cant see if the fix im about to push will actually work, but i can guess that it will, it has no reason not to

**Comment by Flux3on** [comment]
Tbh even +32ms global offset feels really early

**Comment by Beyley** [comment]
> Tbh even +32ms global offset feels really early

There really isnt much you can do about the terrible offset issues, android by default has a huge audio delay that i cant really fix

**Comment by s1rcheese** [comment]
Then is it possible to delay sound by 32 seconds?

**Comment by Beyley** [comment]
> Then is it possible to delay sound by 32 seconds?

i assume you mean 32 miliseconds, in which case yes it is possible, and its in the options

**Comment by s1rcheese** [comment]
I did mean MS, but I was thinking specificlly for Android it would automaticlly be delayed by 32ms. (Or there could be a note that says that 32ms is an issue)

File: osu!stream/Resources/Resource.designer.cs [hunk 2]
```diff
@@ -0,0 +1,8698 @@
...
+//------------------------------------------------------------------------------
+// <auto-generated>
+//     This code was generated by a tool.
+//
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+//------------------------------------------------------------------------------
+
+[assembly: global::Android.Runtime.ResourceDesignerAttribute("osum.Resource", IsApplication=true)]
```
Comment (peppy) [inline osu!stream/Resources/Resource.designer.cs:11]: Is this file actually required? If it can be removed, it would aid in me reviewing this PR more easily.

Comment (Beyley) [reply osu!stream/Resources/Resource.designer.cs]: > If it is required, at least try and reduce the number of lines. Right now it looks to contain a lot of resources which don't need to be used by the game.

The file is regenerated by every build, so i added it to the .gitignore

**Comment by peppy** [comment]
@Beyley finally got around to testing this (sorry for the delay) and now have an error log:

```csharp
Forwarding debugger port 8900
Detecting existing process
> am start -n "sh.ppy.osustream/crc6484b8820ed7fe2f2b.Application"
> Starting: Intent { cmp=sh.ppy.osustream/crc6484b8820ed7fe2f2b.Application launchParam=MultiScreenLaunchParams { mDisplayId=0 mFlags=0 } }

[Zygote] v2
[libpersona] KNOX_SDCARD checking this for 10214
[libpersona] KNOX_SDCARD not a persona
[Zygote] accessInfo : 0
[SELinux] SELinux selinux_android_compute_policy_index : Policy Index[2],  Con:u:r:zygote:s0 RAM:SEPF_SECMOBILE_7.0_0009, [-1 -1 -1 -1 0 1]
[SELinux] SELinux: seapp_context_lookup: seinfo=untrusted, level=s0:c512,c768, pkgname=sh.ppy.osustream 
[TimaKeyStoreProvider] TimaKeyStore is not enabled: cannot add TimaSignature Service and generateKeyPair Service
[monodroid-gc] GREF GC Threshold: 46080
[art] Before Android 4.1, method android.graphics.PorterDuffColorFilter android.support.graphics.drawable.VectorDrawableCompat.updateTintFilter(android.graphics.PorterDuffColorFilter, android.content.res.ColorStateList, android.graphics.PorterDuff$Mode) would have incorrectly overridden the package-private method in android.graphics.drawable.Drawable
[AndroidRuntime] Shutting down VM
[MonoDroid] UNHANDLED EXCEPTION:
[MonoDroid] Java.Lang.LinkageError: no "I" field "layoutInDisplayCutoutMode" in class "Landroid/view/WindowManager$LayoutParams;" or its superclasses
[MonoDroid]   at Java.Interop.JniEnvironment+InstanceFields.GetFieldID (Java.Interop.JniObjectReference type, System.String name, System.String signature) [0x0005b] in <4cc641ad379b4111af2059fd28d44558>:0 
[MonoDroid]   at Java.Interop.JniType.GetInstanceField (System.String name, System.String signature) [0x0000c] in <4cc641ad379b4111af2059fd28d44558>:0 
[MonoDroid]   at Java.Interop.JniPeerMembers+JniInstanceFields.GetFieldInfo (System.String encodedMember) [0x00036] in <4cc641ad379b4111af2059fd28d44558>:0 
[MonoDroid]   at Java.Interop.JniPeerMembers+JniInstanceFields.SetValue (System.String encodedMember, Java.Interop.IJavaPeerable self, System.Int32 value) [0x00006] in <4cc641ad379b4111af2059fd28d44558>:0 
[MonoDroid]   at Android.Views.WindowManagerLayoutParams.set_LayoutInDisplayCutoutMode (Android.Views.LayoutInDisplayCutoutMode value) [0x0000a] in <b3c4d9bb8aa64cbc80ab38b5ea9ef86b>:0 
[MonoDroid]   at osum.Application.OnCreate (Android.OS.Bundle savedInstanceState) [0x00030] in <aaedd9a2fcd34881be900dcf86c0c049>:0 
[MonoDroid]   at Android.App.Activity.n_OnCreate_Landroid_os_Bundle_ (System.IntPtr jnienv, System.IntPtr native__this, System.IntPtr native_savedInstanceState) [0x0000f] in <b3c4d9bb8aa64cbc80ab38b5ea9ef86b>:0 
[MonoDroid]   at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.2(intptr,intptr,intptr)
[MonoDroid]   --- End of managed Java.Lang.LinkageError stack trace ---
[MonoDroid] java.lang.NoSuchFieldError: no "I" field "layoutInDisplayCutoutMode" in class "Landroid/view/WindowManager$LayoutParams;" or its superclasses
[MonoDroid] 	at crc6484b8820ed7fe2f2b.Application.n_onCreate(Native Method)
[MonoDroid] 	at crc6484b8820ed7fe2f2b.Application.onCreate(Application.java:43)
[MonoDroid] 	at android.app.Activity.performCreate(Activity.java:6942)
[MonoDroid] 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1126)
[MonoDroid] 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2880)
[MonoDroid] 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2988)
[MonoDroid] 	at android.app.ActivityThread.-wrap14(ActivityThread.java)
[MonoDroid] 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1631)
[MonoDroid] 	at android.os.Handler.dispatchMessage(Handler.java:102)
[MonoDroid] 	at android.os.Looper.loop(Looper.java:154)
[MonoDroid] 	at android.app.ActivityThread.main(ActivityThread.java:6682)
[MonoDroid] 	at java.lang.reflect.Method.invoke(Native Method)
[MonoDroid] 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1520)
[MonoDroid] 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1410)
[MonoDroid] 
[AndroidRuntime] FATAL EXCEPTION: main
[AndroidRuntime] Process: sh.ppy.osustream, PID: 7370
[AndroidRuntime] java.lang.NoSuchFieldError: no "I" field "layoutInDisplayCutoutMode" in class "Landroid/view/WindowManager$LayoutParams;" or its superclasses
[AndroidRuntime] 	at crc6484b8820ed7fe2f2b.Application.n_onCreate(Native Method)
[AndroidRuntime] 	at crc6484b8820ed7fe2f2b.Application.onCreate(Application.java:43)
[AndroidRuntime] 	at android.app.Activity.performCreate(Activity.java:6942)
[AndroidRuntime] 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1126)
[AndroidRuntime] 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2880)
[AndroidRuntime] 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2988)
[AndroidRuntime] 	at android.app.ActivityThread.-wrap14(ActivityThread.java)
[AndroidRuntime] 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1631)
[AndroidRuntime] 	at android.os.Handler.dispatchMessage(Handler.java:102)
[AndroidRuntime] 	at android.os.Looper.loop(Looper.java:154)
[AndroidRuntime] 	at android.app.ActivityThread.main(ActivityThread.java:6682)
[AndroidRuntime] 	at java.lang.reflect.Method.invoke(Native Method)
[AndroidRuntime] 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1520)
[AndroidRuntime] 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1410)


```

Looks like something may need to be changed in the resource setup to support my device. I'll leave it in your hands, let me know if you need further testing.

**Comment by Beyley** [comment]
> @Beyley finally got around to testing this (sorry for the delay) and now have an error log:
> 
> ```cs
> Forwarding debugger port 8900
> Detecting existing process
> > am start -n "sh.ppy.osustream/crc6484b8820ed7fe2f2b.Application"
> > Starting: Intent { cmp=sh.ppy.osustream/crc6484b8820ed7fe2f2b.Application launchParam=MultiScreenLaunchParams { mDisplayId=0 mFlags=0 } }
> 
> [Zygote] v2
> [libpersona] KNOX_SDCARD checking this for 10214
> [libpersona] KNOX_SDCARD not a persona
> [Zygote] accessInfo : 0
> [SELinux] SELinux selinux_android_compute_policy_index : Policy Index[2],  Con:u:r:zygote:s0 RAM:SEPF_SECMOBILE_7.0_0009, [-1 -1 -1 -1 0 1]
> [SELinux] SELinux: seapp_context_lookup: seinfo=untrusted, level=s0:c512,c768, pkgname=sh.ppy.osustream 
> [TimaKeyStoreProvider] TimaKeyStore is not enabled: cannot add TimaSignature Service and generateKeyPair Service
> [monodroid-gc] GREF GC Threshold: 46080
> [art] Before Android 4.1, method android.graphics.PorterDuffColorFilter android.support.graphics.drawable.VectorDrawableCompat.updateTintFilter(android.graphics.PorterDuffColorFilter, android.content.res.ColorStateList, android.graphics.PorterDuff$Mode) would have incorrectly overridden the package-private method in android.graphics.drawable.Drawable
> [AndroidRuntime] Shutting down VM
> [MonoDroid] UNHANDLED EXCEPTION:
> [MonoDroid] Java.Lang.LinkageError: no "I" field "layoutInDisplayCutoutMode" in class "Landroid/view/WindowManager$LayoutParams;" or its superclasses
> [MonoDroid]   at Java.Interop.JniEnvironment+InstanceFields.GetFieldID (Java.Interop.JniObjectReference type, System.String name, System.String signature) [0x0005b] in <4cc641ad379b4111af2059fd28d44558>:0 
> [MonoDroid]   at Java.Interop.JniType.GetInstanceField (System.String name, System.String signature) [0x0000c] in <4cc641ad379b4111af2059fd28d44558>:0 
> [MonoDroid]   at Java.Interop.JniPeerMembers+JniInstanceFields.GetFieldInfo (System.String encodedMember) [0x00036] in <4cc641ad379b4111af2059fd28d44558>:0 
> [MonoDroid]   at Java.Interop.JniPeerMembers+JniInstanceFields.SetValue (System.String encodedMember, Java.Interop.IJavaPeerable self, System.Int32 value) [0x00006] in <4cc641ad379b4111af2059fd28d44558>:0 
> [MonoDroid]   at Android.Views.WindowManagerLayoutParams.set_LayoutInDisplayCutoutMode (Android.Views.LayoutInDisplayCutoutMode value) [0x0000a] in <b3c4d9bb8aa64cbc80ab38b5ea9ef86b>:0 
> [MonoDroid]   at osum.Application.OnCreate (Android.OS.Bundle savedInstanceState) [0x00030] in <aaedd9a2fcd34881be900dcf86c0c049>:0 
> [MonoDroid]   at Android.App.Activity.n_OnCreate_Landroid_os_Bundle_ (System.IntPtr jnienv, System.IntPtr native__this, System.IntPtr native_savedInstanceState) [0x0000f] in <b3c4d9bb8aa64cbc80ab38b5ea9ef86b>:0 
> [MonoDroid]   at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.2(intptr,intptr,intptr)
> [MonoDroid]   --- End of managed Java.Lang.LinkageError stack trace ---
> [MonoDroid] java.lang.NoSuchFieldError: no "I" field "layoutInDisplayCutoutMode" in class "Landroid/view/WindowManager$LayoutParams;" or its superclasses
> [MonoDroid] 	at crc6484b8820ed7fe2f2b.Application.n_onCreate(Native Method)
> [MonoDroid] 	at crc6484b8820ed7fe2f2b.Application.onCreate(Application.java:43)
> [MonoDroid] 	at android.app.Activity.performCreate(Activity.java:6942)
> [MonoDroid] 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1126)
> [MonoDroid] 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2880)
> [MonoDroid] 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2988)
> [MonoDroid] 	at android.app.ActivityThread.-wrap14(ActivityThread.java)
> [MonoDroid] 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1631)
> [MonoDroid] 	at android.os.Handler.dispatchMessage(Handler.java:102)
> [MonoDroid] 	at android.os.Looper.loop(Looper.java:154)
> [MonoDroid] 	at android.app.ActivityThread.main(ActivityThread.java:6682)
> [MonoDroid] 	at java.lang.reflect.Method.invoke(Native Method)
> [MonoDroid] 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1520)
> [MonoDroid] 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1410)
> [MonoDroid] 
> [AndroidRuntime] FATAL EXCEPTION: main
> [AndroidRuntime] Process: sh.ppy.osustream, PID: 7370
> [AndroidRuntime] java.lang.NoSuchFieldError: no "I" field "layoutInDisplayCutoutMode" in class "Landroid/view/WindowManager$LayoutParams;" or its superclasses
> [AndroidRuntime] 	at crc6484b8820ed7fe2f2b.Application.n_onCreate(Native Method)
> [AndroidRuntime] 	at crc6484b8820ed7fe2f2b.Application.onCreate(Application.java:43)
> [AndroidRuntime] 	at android.app.Activity.performCreate(Activity.java:6942)
> [AndroidRuntime] 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1126)
> [AndroidRuntime] 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2880)
> [AndroidRuntime] 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2988)
> [AndroidRuntime] 	at android.app.ActivityThread.-wrap14(ActivityThread.java)
> [AndroidRuntime] 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1631)
> [AndroidRuntime] 	at android.os.Handler.dispatchMessage(Handler.java:102)
> [AndroidRuntime] 	at android.os.Looper.loop(Looper.java:154)
> [AndroidRuntime] 	at android.app.ActivityThread.main(ActivityThread.java:6682)
> [AndroidRuntime] 	at java.lang.reflect.Method.invoke(Native Method)
> [AndroidRuntime] 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1520)
> [AndroidRuntime] 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1410)
> ```
> 
> Looks like something may need to be changed in the resource setup to support my device. I'll leave it in your hands, let me know if you need further testing.

Thanks for testing it out, could i get some details of your device and how you launched it? i usually just sideload the APK
The commit i just pushed *should* fix that crash, but not 100% as i dont have an old enough device

**Comment by peppy** [comment]
Thanks, it's working great now!

One thing I noticed is that if you background the application while playing a beatmap, it doesn't seem to go to the pause menu like it does on iOS. Then, on resuming you need to manually pause then resume to actually resume gameplay.

Not sure how simple that is to fix.

**Comment by Beyley** [comment]
> Thanks, it's working great now!
> 
> One thing I noticed is that if you background the application while playing a beatmap, it doesn't seem to go to the pause menu like it does on iOS. Then, on resuming you need to manually pause then resume to actually resume gameplay.
> 
> Not sure how simple that is to fix.

That shouldnt be too hard of a fix

> 

Should be fixed

**Comment by peppy** [comment]
Seems like I can't push changes to your branch. If possible can you check the checkbox on the right that allows me to do that?

**Comment by Beyley** [comment]
> Seems like I can't push changes to your branch. If possible can you check the checkbox on the right that allows me to do that?


58UfUk
seems to already be done, should i try manually adding you as a collaborator on my repo?

**Comment by peppy** [comment]
That might be best, I think it may have something to do with the fork chain (ie. you forked from someone else rather than this fork directly)

**Comment by Beyley** [comment]
> That might be best, I think it may have something to do with the fork chain (ie. you forked from someone else rather than this fork directly)

hopefully that worked, ive never had good experiences with githubs permission and collaborator system

**Comment by peppy** [comment]
Did indeed work, thanks!

Going to go through and do some code quality changes on this, then test playing over the weekend. Looking like we're in a pretty good state though!

**Comment by peppy** [comment]
Also looks like the app doesn't stop playing main menu music when going back to the home screen, is that just me?

**Comment by Beyley** [comment]
> Also looks like the app doesn't stop playing main menu music when going back to the home screen, is that just me?

i guess i entirely forgot about that, i was planning on adding media controls so you could listen to the music you downloaded and such, but now that i think about it, it would require a ton more changes then i originally thought, so ill just pause it for now

File: osu!stream/Support/Android/GameBaseAndroid.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+            NativeAssetManagerAndroid.Manager = activity.Assets;
+        }
+
+        public override void Run()
+        {
+            // Before we run anything, let's check if this is the first run...
+            // If this is the first run, let's import the packaged beatmaps from our assets!
+            // This is unfortunately required because Android sucks, thanks Google.
+            if (Config.GetValue("firstrun", true))
```
Comment (peppy) [inline osu!stream/Support/Android/GameBaseAndroid.cs:33]: Not 100% sure on this, but I guess we'll go with it for now. Will revisit before potentially releasing to the play store.

**Review by peppy — APPROVED** [review]

**Comment by peppy** [comment]
Thanks to those that made this happen. There's still some clean-up I want to do on the code, but this is working and working quite well. Note that the offsets being used are those made for iOS and as such definitely need some tweaking.

It'd be a waste not to push this out to the google play store, so I will aim to do so sometime in 2022.
